docs/usage/jssengine.md

```diff
@@ -0,0 +1,521 @@
+```
+
+will enable TLSv1.1, TLSv1.2, and TLSv1.3.
+
+Alternative methods are available that take JSS standard `SSLVersion` and
+`SSLVersionRange` values as parameters; see the `JSSEngine` javadoc for
+more information.
+
+#### Choosing Cipher Suite
+
+Configuring cipher suites is performed using the standard `SSLEngine`
+configuration method of passing a list of cipher suites to
+`setEnabledCipherSuites`. We filter the list of passed cipher suites to
+only those allowed by local policy. For example:
+
+```java
+// SSLEngine inst;
+inst.setEnabledCipherSuites(new String[] { "TLS_AES_128_GCM_SHA256" });
+```
+
+will enable only a single TLSv1.3 cipher suite.
+
+Alternative methods are available that take JSS standard `SSLCipher`
+values as parameters; see the `JSSEngine` javadoc for more information.
+
+#### Using `JSSParameters`
+
+`JSSParameters` largely aligns with `SSLParameters` except that it allows
+two important introductions:
+
+ 1. Selection of key material, like above. See the javadocs on `JSSParameters`
+    for more information.
+ 2. Setting the peer's hostname, for use with validation of certificates. This
+    allows us to tie into NSS's hostname verification directly, instead of
+    responding after the fact by closing the connection.
+
+#### Session Control
+
+The `JSSEngine` lacks many of the session control functions other `SSLEngine`
+implementations might have. In particular, we:
+
+ - Always enable session resumption; this cannot be disabled.
+ - Allow forced expiration of a session as long as the `SSLEngine`'s
+   connection isn't yet closed.
+ - Report accurate creation/expiration/last accessed times.
+
+However, other features of sessions (such as configuring location and size of
+the session cache) aren't yet configurable.
+
+
+## Design of the `JSSEngine`
+
+### Class Structuring
+
+The below is a digram showing the structure of `JSSEngine` classes:
+
+                           -----------
+                          | JSSEngine |-------------------------
+                           -----------                          \
+                            /       \                            \
+     ------------------------       ------------------------     ------
+    | JSSEngineReferenceImpl |     | JSSEngineOptimizedImpl |   | .... |
+     ------------------------       ------------------------     ------
+
+`JSSEngine` is an abstract class extending [`SSLEngine`][javax.ssl-engine].
+This class implements some of the boilerplate required for implementing a
+`SSLEngine`, including handling cipher and protocol version configuration.
+Individual implementations implement `wrap`, `unwrap`, and whatever specifics
+are necessary to initialize and release the SSL-backed `PRFileDesc`.
+
+We expect two primary implementations:
+
+ - `JSSEngineReferenceImpl`, a reference implementation with more logging
+   and debugging statements. This also includes port-based debugging, so
```

Comment:
Working on the following sentence seems a bit off to me maybe just a typo.

Reply:
"is without writing to the network"

Reply:
Yeah, good catch. Thanks. :)